crates/common/src/ens.rs

```diff
@@ -0,0 +1,149 @@
+//! ENS Name resolving utilities.
+#![allow(missing_docs)]
+use std::{str::FromStr, sync::Arc};
+
+use alloy_primitives::{address, keccak256, Address, B256};
+use alloy_providers::provider::TempProvider;
+use alloy_sol_types::sol;
+
+// ENS Registry and Resolver contracts.
+sol! {
+    #[sol(rpc)]
+    // ENS Registry contract.
+    contract EnsRegistry {
+        /// Returns the resolver for the specified node.
+        function resolver(bytes32 node) view returns (address);
+    }
+
+    #[sol(rpc)]
+    // ENS Resolver interface.
+    contract EnsResolver {
+        // Returns the address associated with the specified node.
+        function addr(bytes32 node) view returns (address);
+    }
+}
+
+/// ENS registry address (`0x00000000000C2E074eC69A0dFb2997BA6C7d2e1e`)
+pub const ENS_ADDRESS: Address = address!("00000000000C2E074eC69A0dFb2997BA6C7d2e1e");
+
+/// ENS name or Ethereum Address.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum NameOrAddress {
+    /// An ENS Name (format does not get checked)
+    Name(String),
+    /// An Ethereum Address
+    Address(Address),
+}
+
+impl NameOrAddress {
+    /// Resolves the name to an Ethereum Address.
+    pub async fn resolve<P: TempProvider>(
```

Comment:
should this be an extension crate in alloy?

Reply:
Yeah I think we can eventually move this to alloy—I'd like to do this once the provider's a bit more stable though, after this PR

Reply:
This type should be in alloy yes

Reply:
let's do it after https://github.com/alloy-rs/alloy/pull/190 is merged, else we'll need to refactor twice